### PR TITLE
Fix bugs on windows

### PR DESCRIPTION
**Pull request by TeBoring**

### CODE REVIEW

**Comment by TeBoring** [comment]
@xfxyjwf

File: src/google/protobuf/compiler/cpp/cpp_unittest.cc [hunk 1]
```diff
@@ -55,7 +55,9 @@
 #include <google/protobuf/unittest.pb.h>
 #include <google/protobuf/unittest_optimize_for.pb.h>
 #include <google/protobuf/unittest_embed_optimize_for.pb.h>
+#if !defined(_MSC_VER)
 #include <google/protobuf/unittest_enormous_descriptor.pb.h>
```
Comment (xfxyjwf) [inline src/google/protobuf/compiler/cpp/cpp_unittest.cc]: Add some comments on why it's excluded.

File: src/google/protobuf/map.h [hunk 2]
```diff
@@ -853,6 +853,28 @@ struct hash<google::protobuf::MapKey> {
         return 0;
     }
   }
+  bool
```
Comment (xfxyjwf) [inline src/google/protobuf/map.h]: Is this only needed for MSVC build? guard it with a macro?

File: src/google/protobuf/map.h [hunk 3]
```diff
@@ -853,6 +853,28 @@ struct hash<google::protobuf::MapKey> {
...
+  operator()(const google::protobuf::MapKey& map_key1,
+             const google::protobuf::MapKey& map_key2) const {
+    switch (map_key1.type()) {
+#define COMPARE_CPPTYPE(CPPTYPE, CPPTYPE_METHOD)             \
```
Comment (xfxyjwf) [inline src/google/protobuf/map.h]: undef the macro after use.

File: src/google/protobuf/map.h (hunk 2 above)
Comment (TeBoring) [inline src/google/protobuf/map.h]: For any platform that uses map

**Comment by xfxyjwf** [comment]
LGTM

It seems this breaks the travis build on linux. Could you check that?

**Pull request merged**